services/src/server.rs

```diff
@@ -64,57 +51,204 @@ pub async fn start_server(
-        handlers::workflows::get_workflow_metadata_handler(ctx.clone()),
-        handlers::workflows::get_workflow_provenance_handler(ctx.clone()),
-        handlers::session::anonymous_handler(ctx.clone()),
-        handlers::session::session_handler(ctx.clone()),
-        handlers::session::session_project_handler(ctx.clone()),
-        handlers::session::session_view_handler(ctx.clone()),
-        handlers::projects::create_project_handler(ctx.clone()),
-        handlers::projects::list_projects_handler(ctx.clone()),
-        handlers::projects::update_project_handler(ctx.clone()),
-        handlers::projects::delete_project_handler(ctx.clone()),
-        handlers::projects::load_project_handler(ctx.clone()),
-        handlers::datasets::get_dataset_handler(ctx.clone()),
-        handlers::datasets::auto_create_dataset_handler(ctx.clone()),
-        handlers::datasets::create_dataset_handler(ctx.clone()),
-        handlers::datasets::suggest_meta_data_handler(ctx.clone()),
-        handlers::datasets::list_providers_handler(ctx.clone()),
-        handlers::datasets::list_external_datasets_handler(ctx.clone()),
-        handlers::datasets::list_datasets_handler(ctx.clone()), // must come after `list_external_datasets_handler`
-        handlers::wcs::wcs_handler(ctx.clone()),
-        handlers::wms::wms_handler(ctx.clone()),
-        handlers::wfs::wfs_handler(ctx.clone()),
-        handlers::plots::get_plot_handler(ctx.clone()),
-        handlers::upload::upload_handler(ctx.clone()),
-        handlers::spatial_references::get_spatial_reference_specification_handler(ctx.clone()),
-        show_version_handler(), // TODO: allow disabling this function via config or feature flag
-        serve_static_directory(static_files_dir)
-    )
-    .recover(handle_rejection);
+    let wrapped_ctx = web::Data::new(ctx);
+
+    HttpServer::new(move || {
+        let app = App::new()
+            .app_data(wrapped_ctx.clone())
+            .wrap(middleware::Logger::default())
+            .wrap(middleware::NormalizePath::default())
+            .configure(configure_extractors)
+            .configure(init_routes::<C>);
 
-    let task = if let Some(receiver) = shutdown_rx {
-        let (_, server) = warp::serve(handler).bind_with_graceful_shutdown(bind_address, async {
-            receiver.await.ok();
-        });
-        tokio::task::spawn(server)
-    } else {
-        let server = warp::serve(handler).bind(bind_address);
-        tokio::task::spawn(server)
-    };
+        if let Some(static_files_dir) = static_files_dir.clone() {
+            app.service(Files::new("/static", static_files_dir))
+        } else {
+            app
+        }
+    })
+    .bind(bind_address)?
+    .run()
+    .await
+    .map_err(Into::into)
+}
+
+pub(crate) fn configure_extractors(cfg: &mut web::ServiceConfig) {
+    cfg.app_data(web::JsonConfig::default().error_handler(|err, _req| {
+        match err {
+            JsonPayloadError::Overflow => todo!(),
```

Comment:
What is this and could we throw an error instead?
`todo!` panics and would kill our threads.

Reply:
06a84c4